File: src/apis/callRecordings.js
const express = require('express');
const axios = require('axios');
const jwt = require('jsonwebtoken');

const router = express.Router();

// Middleware for token-based authentication
router.use((req, res, next) => {
    const token = req.headers['authorization']?.split(' ')[1];
    if (!token) return res.status(403).send('Forbidden - No token provided');
    jwt.verify(token, 'your_secret_key', (err, decoded) => {
        if (err) return res.status(403).send('Forbidden - Invalid token');
        req.user = decoded; // store user info
        next();
    });
});

// Endpoint to fetch call recordings
router.get('/call-recordings', async (req, res) => {
    try {
        // Fetch call recordings from external API
        const response = await axios.get('https://externalapi.com/call-recordings', {
            headers: {
                'Authorization': `Bearer ${req.headers['authorization'].split(' ')[1]}`
            }
        });

        const callRecordingsData = response.data;
        const callRecordings = [];
        const fundingInfo = [];

        // Parse response for call recordings and funding information
        callRecordingsData.forEach(recording => {
            callRecordings.push({ id: recording.id, title: recording.title });
            fundingInfo.push({ recordingId: recording.id, funding: recording.funding });
        });

        // Send response with separate entities for call recordings and funding info
        res.json({ callRecordings, fundingInfo });
    } catch (error) {
        console.error('Error fetching call recordings:', error);
        res.status(500).send('Internal Server Error');
    }
});

module.exports = router;

File: api/call_recordings.go
package api

import (
    "net/http"
    "encoding/json"
    "github.com/gorilla/mux"
    "github.com/dgrijalva/jwt-go"
)

// CallRecording represents call recording data
type CallRecording struct {
    ID   string `json:"id"`
    URL  string `json:"url"`
}

// FundingInfo represents funding-related data
type FundingInfo struct {
    RecordingID string  `json:"recording_id"`
    Amount      float64 `json:"amount"`
    Source      string  `json:"source"`
}

// Authentication middleware
func AuthenticationMiddleware(next http.Handler) http.Handler {
    return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
        tokenString := r.Header.Get("Authorization")
        if tokenString == "" {
            http.Error(w, "Forbidden", http.StatusForbidden)
            return
        }
        // Validate token (simplified for demo purposes)
        _, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
            // Check token signing method etc. (not implemented here)
            return []byte("your-secret-key"), nil
        })
        if err != nil {
            http.Error(w, "Forbidden", http.StatusForbidden)
            return
        }
        next.ServeHTTP(w, r)
    })
}

// Fetch call recordings and funding information
func FetchCallRecordings(w http.ResponseWriter, r *http.Request) {
    // Simulated response from an external API
    callRecordings := []CallRecording{
        {ID: "1", URL: "http://example.com/call1"},
        {ID: "2", URL: "http://example.com/call2"},
    }

    fundingInfos := []FundingInfo{
        {RecordingID: "1", Amount: 1000.00, Source: "Source A"},
        {RecordingID: "2", Amount: 2000.00, Source: "Source B"},
    }

    response := map[string]interface{}{
        "call_recordings": callRecordings,
        "funding_info":    fundingInfos,
    }

    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(response)
}

// Setup routes
func SetupRoutes() {
    r := mux.NewRouter()
    r.Use(AuthenticationMiddleware)
    r.HandleFunc("/api/call_recordings", FetchCallRecordings).Methods("GET")
    http.ListenAndServe(":8080", r)
}